src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,546 @@
+{
+    // Extract command fields from the request.
+    uint16_t sessionId = req.webRTCSessionID;
+    auto sdpSpan       = req.sdp;
+
+    NodeId peerNodeId = GetNodeIdFromCtx(ctx.mCommandHandler);
+
+    WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+    // Respond with NOT_FOUND if the session is not valid.
+    if (!existingSession)
+    {
+        ChipLogError(Zcl, "HandleProvideAnswer: No existing session with ID=%u.", sessionId);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::NotFound);
+        return;
+    }
+
+    // Verify that the session's PeerNodeID and FabricIndex match those from the secure session context.
+    if (peerNodeId != existingSession->peerNodeID)
+    {
+        ChipLogError(Zcl, "HandleProvideAnswer: Session ID=%u does not match PeerNodeID or FabricIndex.", sessionId);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::NotFound);
+        return;
+    }
+
+    std::string sdpAnswer(sdpSpan.data(), sdpSpan.size());
+    ctx.mCommandHandler.AddStatus(
+        ctx.mRequestPath, Protocols::InteractionModel::ClusterStatusCode(mDelegate.HandleProvideAnswer(sessionId, sdpAnswer)));
+}
+
+void WebRTCTransportProviderServer::HandleProvideICECandidates(HandlerContext & ctx,
+                                                               const Commands::ProvideICECandidates::DecodableType & req)
+{
+    // Extract command fields from the request.
+    uint16_t sessionId = req.webRTCSessionID;
+
+    NodeId peerNodeId = GetNodeIdFromCtx(ctx.mCommandHandler);
+
+    std::vector<std::string> candidates;
+    auto iter = req.ICECandidates.begin();
+    while (iter.Next())
+    {
+        // Get current candidate CharSpan and convert to std::string.
+        const CharSpan & candidateSpan = iter.GetValue();
+        candidates.emplace_back(candidateSpan.data(), candidateSpan.size());
+    }
+
+    // Check the validity of the list.
+    CHIP_ERROR listErr = iter.GetStatus();
+    if (listErr != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "HandleProvideICECandidates: ICECandidates list error: %s", chip::ErrorStr(listErr));
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::ConstraintError);
+        return;
+    }
+
+    WebRTCSessionStruct * existingSession = FindSession(sessionId);
+
+    // If WebRTCSessionID is not valid, respond with NOT_FOUND.
+    if (!existingSession)
+    {
+        ChipLogError(Zcl, "HandleProvideICECandidates: No existing session with ID=%u.", sessionId);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::NotFound);
+        return;
+    }
+
+    // Verify that the session belongs to the accessing peer.
+    // If it doesn't match, respond with NOT_FOUND.
+    if (peerNodeId != existingSession->peerNodeID)
```

Comment:
Again: this duplicated "is this the right session" code should be factored out.  Followup is OK for that, but make sure it's tracked and has TODO comments pointing to the issue.